src/keys.rs

```diff
@@ -1,70 +1,193 @@
 #[derive(Debug)]
-pub(crate) struct NullifierDerivingKey;
+pub(crate) struct NullifierDerivingKey(pallas::Base);
 
 impl From<&SpendingKey> for NullifierDerivingKey {
-    fn from(_: &SpendingKey) -> Self {
-        todo!()
+    fn from(sk: &SpendingKey) -> Self {
+        NullifierDerivingKey(to_base(prf_expand(&sk.0, &[0x07])))
     }
 }
 
 /// A key that provides the capability to view incoming and outgoing transactions.
 ///
 /// This key is useful anywhere you need to maintain accurate balance, but do not want the
 /// ability to spend funds (such as a view-only wallet).
-///
-/// TODO: Should we just define the FVK to include extended stuff like the diversifier key?
 #[derive(Debug)]
 pub struct FullViewingKey {
     ak: AuthorizingKey,
     nk: NullifierDerivingKey,
-    rivk: (),
+    rivk: pallas::Scalar,
 }
 
 impl From<&SpendingKey> for FullViewingKey {
-    fn from(_: &SpendingKey) -> Self {
-        todo!()
+    fn from(sk: &SpendingKey) -> Self {
+        FullViewingKey {
+            ak: (&SpendAuthorizingKey::from(sk)).into(),
+            nk: sk.into(),
+            rivk: to_scalar(prf_expand(&sk.0, &[0x08])),
+        }
     }
 }
 
 impl FullViewingKey {
+    /// Defined in [Zcash Protocol Spec § 4.2.3: Orchard Key Components][§4.2.3].
+    ///
+    /// [§4.2.3]: https://zips.z.cash/protocol/nu5.pdf#orchardkeycomponents
+    fn derive_dk_ovk(&self) -> (DiversifierKey, OutgoingViewingKey) {
+        let k = self.rivk.to_bytes();
+        let b = [(&self.ak.0).into(), self.nk.0.to_bytes()];
+        let r = prf_expand_vec(&k, &[&[0x82], &b[0][..], &b[1][..]]);
+        (
+            DiversifierKey(r.as_bytes()[..32].try_into().unwrap()),
+            OutgoingViewingKey(r.as_bytes()[32..].try_into().unwrap()),
+        )
+    }
+
+    /// Returns the default payment address for this key.
+    pub fn default_address(&self) -> Address {
+        self.address(DiversifierKey::from(self).default_diversifier())
+    }
+
     /// Returns the payment address for this key corresponding to the given diversifier.
     pub fn address(&self, d: Diversifier) -> Address {
         IncomingViewingKey::from(self).address(d)
     }
 }
 
+/// A key that provides the capability to derive a sequence of diversifiers.
+#[derive(Debug)]
+pub struct DiversifierKey([u8; 32]);
+
+impl From<&FullViewingKey> for DiversifierKey {
+    fn from(fvk: &FullViewingKey) -> Self {
+        fvk.derive_dk_ovk().0
+    }
+}
+
+/// The index for a particular diversifier.
+#[derive(Clone, Copy, Debug)]
+pub struct DiversifierIndex([u8; 11]);
+
+macro_rules! di_from {
+    ($n:ident) => {
+        impl From<$n> for DiversifierIndex {
+            fn from(j: $n) -> Self {
+                let mut j_bytes = [0; 11];
+                j_bytes[..mem::size_of::<$n>()].copy_from_slice(&j.to_le_bytes());
+                DiversifierIndex(j_bytes)
+            }
+        }
+    };
+}
+di_from!(u8);
+di_from!(u16);
+di_from!(u32);
+di_from!(u64);
+di_from!(usize);
```

Comment:
Are these all expected to be needed?

Reply:
`u8` and `u16` are probably the least likely to be used. `usize` makes sense from a Rust perspective, and `u32, u64` are quite common to see as e.g. customer IDs in SQL tables.

Reply:
Would it make sense to also support u128 and/or UUID (16 octets) as a source directly? If you want to support database identifiers, UUIDs are becoming pretty common.

Reply:
These would require a `TryFrom` implementation (because they might be out-of-range), so I'm less inclined to support them directly; in that case the user probably needs to define an explicit mapping from the UUID to a diversifier (since the UUID is likely random, and thus almost guaranteed to be out-of-range).